Support scalar functions with `CREATE FUNCTION`

lnkuiper opened the pull request:
Implementation of issue #1064 that allows for creation of scalar functions (a.k.a. "macro's") from SQL, with an alias `CREATE MACRO`.

This is made possible by unfolding the original expression, and replacing the parameters with the supplied arguments.
For example:
```
D CREATE FUNCTION add(a,b) AS a+b;
D SELECT add(40,2);
┌────────────┐
│ add(40, 2) │
├────────────┤
│ 42         │
└────────────┘
```

Subqueries, CTE's and nested macro's are supported:

```
D CREATE MACRO showing_off(a,b) AS (WITH cte AS (SELECT a * 2 AS c) SELECT add(cte.c, sq.d) FROM cte, (SELECT b * 3 AS d) AS sq);
D SELECT showing_off(3,4);
┌───────────────────┐
│ showing_off(3, 4) │
├───────────────────┤
│ 18                │
└───────────────────┘

```

There are some limitations, and some things to look out for. Arguments with side-effects are (not yet) supported:

```
D SELECT add(40,random());
Error: Binder Error: Arguments with side-effects are not supported ('random()' was supplied). As a workaround, try creating a CTE that evaluates the argument with side-effects.
D WITH cte AS (SELECT random() AS r) SELECT add(1,r) FROM cte;
┌────────────────────┐
│     add(1, r)      │
├────────────────────┤
│ 1.0317918057098674 │
└────────────────────┘
```

The binder cannot know if a column or parameter is referenced when they have the same name. In case of conflicting names, parameters always get priority. To work around this, the table name must be supplied to disambiguate:
```
D CREATE TABLE integers (i INT);
D INSERT INTO integers VALUES (1);
D CREATE MACRO ambiguous(i) AS (SELECT i + i FROM integers);
D SELECT ambiguous(5);
┌──────────────┐
│ ambiguous(5) │
├──────────────┤
│ 10           │
└──────────────┘
D CREATE MACRO unambiguous AS (SELECT integers.i + i FROM integers);
D CREATE MACRO unambiguous(i) AS (SELECT integers.i + i FROM integers);
D SELECT unambiguous(5);
┌────────────────┐
│ unambiguous(5) │
├────────────────┤
│ 6              │
└────────────────┘

```

Macro's are bound at creation time for verification, meaning that `CREATE MACRO add(a, b) AS a + b + c;` is not possible.

Macro's are stored in the catalog of a schema, and are persistent.
```
D CREATE SCHEMA macros;
D CREATE MACRO macros.add(a,b);
D SELECT macros.add(40,2);
┌────────────┐
│ add(40, 2) │
├────────────┤
│ 42         │
└────────────┘
```
Because macro's were added to storage, backwards compatibility with storage was broken


This functionality will be used to implement #952.

Feel free to test it out, and I am happy to receive any feedback!

Commit:
init

Commit:
add CREATE FUNCTION to grammar

Commit:
add name and fix numbering

Commit:
re-generate after adding CREATE FUNCTION

Commit:
create function statement

Commit:
init

Commit:
add create function to transformer and transform it to info

Commit:
properly typedef CreateFunctionStmt

Commit:
bind create function *WIP*

Commit:
add CREATE FUNCTION catalog entry, get schema, re-generate grammar

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
Add MACRO alias for CREATE FUNCTION, trying to process ParsedExpression *WIP*

Commit:
move BindCreateFunction to separate cpp file

Commit:
function arguments as dummy NULL columns in binder

Commit:
simplify CREATE FUNCTION and create a physical operator

Commit:
add macro function entry to switches

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
bind macro function call *WIP*

Commit:
create macro function class to store ParsedExpressions in

Commit:
add unique_ptr parsed expression iterator

Commit:
properly bind macro function and add it to more switches

Commit:
merge master and make format

Commit:
switch to new assert

Commit:
spaces to tabs

Commit:
macro expression executor

Commit:
merge master CREATE FUNCTION

Commit:
merge master CREATE FUNCTION

Commit:
remove BoundMacroExpression (not needed)

Commit:
add possibility empty function arg list

Commit:
support nested macro's *WIP*

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
clean up CREATE FUNCTION yacc file

Commit:
bind macro parameters & args

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
cleanup macro_binding and properly bind non-macro columns

Commit:
better naming

Commit:
only bind table name if col is definitely not a macro parameter

Commit:
push to desktop *WIP*

Commit:
consistent macro naming

Commit:
clean up create function grammar

Commit:
fix wrong test

Commit:
throw error on incorrect macro signature

Commit:
properly check for arguments with side-effects and pretty format error

Commit:
make CI happy

Commit:
more macro tests

Commit:
push to PC *WIP*

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
avoid nullptr access

Commit:
unfold macro expression recursively *WIP*

Commit:
macro unfolding

Commit:
fix recursive macro unfolding

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
allow parameters in macro CTE's

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
copy the alias of macro parameters

Commit:
copy the alias of macro parameters

Commit:
replace macro params in tableref

Commit:
move binding of macro's to separate file

Commit:
allow dropping of macro's

Commit:
support recursive CTE and setop nodes in macro

Commit:
persist macro's

Commit:
add test for persistent macro's

Commit:
undo 'make format'

Commit:
consistent MACRO naming

Commit:
undo 'make format'

Commit:
additional test case

Commit:
consistent 'macro' naming

Commit:
make github actions happy

Commit:
increment storage number for persistent macro's

Commit:
generate storage version after incrementing the number (oops)

Commit:
properly update storage version for persistent macro's

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
add (and fix) test case

Commit:
add schema macro test case

Commit:
add cases for TableRefs that do not need unfolding

Mytherin reviewed (commented):
Thanks for the PR! Looks great. I added some (mostly stylistic) comments. Besides, that I would like to see some more tests of the following (correct me if these tests are there, I couldn't see them at first glance):

* Macro with zero parameters
* Macro with name conflict in parameters (i.e. `MACRO(i, i)`)
* Macro with another macro definition inside it (e.g. `add(x,y) AS x+y` -> `double_add(x, y, z) AS add(add(x, y), z)`
* Usage of nested macros (e.g. `add(add(x, y), add(y, z))`)
* Subqueries in macro parameters (e.g. `add((SELECT MIN(i) FROM integers), (SELECT max(i) FROM integers);`)
* Macros within a correlated subquery (e.g. `SELECT (SELECT add(i1.i, i) FROM integers) FROM integers i1`)
* Macros with aggregations (e.g. `MYAVG(x) -> SUM(x) / COUNT(x)`)
* Macros with window functions (e.g. `MYWINDOW(x) -> sum(x) OVER (...)`)
* Macros with prepared statements (e.g.  `prep(x) -> ?+1`) (this should be an error)
* The storage tests I mentioned in the comments below

lnkuiper commented:
Thanks for the feedback! Two of the test cases you mentioned do not work yet, so I still have some work to do :)

Commit:
PR feedback and merge maste

Commit:
rewrite BindExpression api *WIP*

Commit:
change BindExpression API to be able to replace original expression (for MACRO's)

Commit:
use selectbinder for macro creation

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
fix macro storage

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
bump the storage version

Commit:
throw exception on non-foldable limit clause

Commit:
rework parsed expression iterator

Commit:
remove unused import

Commit:
change macro binding to reflect changes in ParsedExpressionIterator

Commit:
fix window macro test

Commit:
test macro's with sequences and recursive CTE's

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

lnkuiper commented:
I've implemented the feedback, and solved the 2 tests that were broken. Additionally, I've added tests with recursive CTE's, sequences, and subqueries. I struggle with writing complex tests because my SQL skills are not that great. However, if you want more tests, don't hesitate to ask! It will be a good learning experience for me.

Could you take a look again? Glad to receive any feedback.

Also, I see this new coverage/coveralls check. Good stuff! Should I aim for a certain %? Or should I aim to cover all branches?

Mytherin commented on src/catalog/catalog_entry/macro_catalog_entry.cpp:
```diff
@@ -0,0 +1,29 @@
+	serializer.WriteString(name);
+    function->expression->Serialize(serializer);
+    serializer.Write<uint32_t>((uint32_t)function->parameters.size());
+	for (auto &param : function->parameters) {
```
There seems to be some weird formatting here, likely a mix of tabs and spaces. Could you run the formatter (`python3 scripts/format.py master`)?

Mytherin commented on src/function/macro_function.cpp:
```diff
@@ -0,0 +1,56 @@
+
+	// check for arguments with side-effects TODO: to support this, a projection must be pushed
+	for (auto &arg : arguments) {
+		if (arg->GetExpressionClass() == ExpressionClass::FUNCTION) {
```
This is actually not entirely correct, since an argument could have hidden side effects (e.g. `i + nextval('seq')`). My CSE optimizer branch (#1143) has a fix for this by adding the `HasSideEffects()` function to expressions, which can be used here. I will merge that one first. Could you then change the side effect detection here to use that function and add a testcase?

lnkuiper replied:
The loop iterates over ParsedExpressions, not Expressions, so the side-effects trick will not work here. I could try to call BindChild to transform it into an Expression, and then call HasSideEffects, not sure if that will work. Will give it a go

Mytherin commented on src/planner/binder/statement/bind_create.cpp:
```diff
@@ -57,6 +61,37 @@ void Binder::BindCreateViewInfo(CreateViewInfo &base) {
+	vector<string> dummy_names;
+	for (idx_t i = 0; i < base.function->parameters.size(); i++) {
+		if (base.function->parameters[i]->expression_class != ExpressionClass::COLUMN_REF)
+			throw BinderException("Invalid parameter \"%s\"", base.function->parameters[i]->ToString());
```
Brackets v2: electric boogaloo

Mytherin commented on src/include/duckdb/planner/binder.hpp:
```diff
@@ -75,6 +76,8 @@ class Binder {
 	//! The alias for the currently processing subquery, if it exists
 	string alias;
+	//! Macro parameter bindings (if any)
+	shared_ptr<MacroBinding> macro_binding = nullptr;
```
Does this need to be a shared pointer? We prefer to avoid shared pointers unless absolutely required (because of the performance penalty, potential leaks because of referential cycles, difficult to reason about ownership). Could this be turned into a unique pointer, or a regular (non-smart) pointer?

lnkuiper replied:
In my first idea of macro bindings, this was needed. However, I've changed the approach since then, and it is no longer needed. A regular pointer will suffice.

Mytherin commented on test/sql/catalog/function/test_simple_macro.test:
```diff
@@ -0,0 +1,150 @@
+statement ok
+CREATE MACRO weird_avg(x) AS (MIN(x) + MAX(x)) / COUNT(x)
+
+# this one is also 21 because there's only 2 integers in there
```
I think this comment is outdated :)

Mytherin reviewed (commented):
Thanks for all the changes, it is starting to look very good! Turned out to be quite a big project!

I have some minor comments, besides those I just thought of one more test case: using a macro inside of a prepared statement, i.e.:

```sql
create function add_one(a) AS a+1;
prepare v1 as select add_one(?::INTEGER);
execute v1(1);
-- 2

-- and with the same prepared statement parameter used multiple times
create function my_square(a) AS a*a;
prepare v2 as select my_square(?::INTEGER);
execute v2(3);
-- 9
```

After that, I think I am all out of ideas for tests :) Excellent job!

Mytherin commented:
> Also, I see this new coverage/coveralls check. Good stuff! Should I aim for a certain %? Or should I aim to cover all branches?

In general we try to cover as much as possible (including any exceptions), but there will always be some paths that are unreachable (we generally throw internal exceptions there - if any internal exception gets thrown it is a bug). It's generally good to check if all the code paths you think you are testing for are actually being taken, though.

One problem we ran into before is that if you are testing for an error using `statement error`, and the query has a typo, the query will fail in the parser and pass the test. However, it will not be testing the actual behavior you want to test. The code coverage helps detect this case. This is something I still want to solve in the tester as well by adding an optional exception message to `statement error` which verifies the *correct* exception is thrown.

Commit:
brackets

Commit:
format

Commit:
regular pointer instead of shared

Commit:
fix outdated comment

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
add arg with side-effects test

Commit:
formatting

Commit:
refactor and clean up macro code

Commit:
Merge remote-tracking branch 'upstream/master' into createfunction

Commit:
clarify exception message

Commit:
formatting

Commit:
copy parameter number on copying parameter expression

Commit:
allow multiple references of same parameter expression

Commit:
add macro_binding to expression_binder

Commit:
allow multiple references to same parameter expression

Commit:
more test cases

Commit:
remove commented code

Commit:
check type consistency when binding parameter expression

Commit:
formatting

lnkuiper commented:
I fixed the tests with prepared statements, and cleaned up the code. When cleaning the code I actually found another bug, which I'm happy to say I fixed. Everything took longer than expected, but I'm happy with the results. Some checks were not successful due to a 500 server error.

I still intend to properly support arguments with side-effects and default parameters, but I'll be glad if this is merged for now.

Thanks for all the help you gave me when implementing this, Mark! And let me know if anything needs changing
